### DISCUSSION TITLE
I can't evaluate on my own datsets.

### DISCUSSION THREAD

**Discussion by JeonghwaYoo-R** [TimeEval/TimeEval#53] [Q&A]
My code is here: 
```
custom_datasets_path = Path("~.json")
dm = DatasetManager(HPI_CLUSTER.akita_dataset_paths[HPI_CLUSTER.BENCHMARK], custom_datasets_file=custom_datasets_path)

# Select datasets and algorithms
datasets = dm.select()
#datasets = datasets[-1:]
# Add algorithms to evaluate...
algorithms = [
    Algorithm(
        name="LOF",
        main=DockerAdapter(
            image_name="registry.gitlab.hpi.de/akita/i/lof",
            tag="latest",  # usually you can use the default here
            skip_pull=True  # set to True because the image is already present from the previous section
        ),
        data_as_file=True,
        training_type=TrainingType.UNSUPERVISED,
        input_dimensionality=InputDimensionality.UNIVARIATE,
    )
]
timeeval = TimeEval(dm, datasets, algorithms,
                    metrics=[DefaultMetrics.ROC_AUC, DefaultMetrics.FIXED_RANGE_PR_AUC])
# execute evaluation
timeeval.run()
# retrieve results
print(timeeval.get_results())
```

My json file about dataset is here: 

> {
>   "dataset_name": {
>     "test_path": "~~.csv"
>   }
> }

Error message is here: 

> hp, ht, pid, tid = _winapi.CreateProcess(executable, args,
> FileNotFoundError: [WinError 2] The system cannot find the file specified
> Evaluating: 100%|██████████| 1/1 [00:00<00:00, 28.83it/s]
> Running FINALIZE phase
> The results contain errors which are filtered out for the final aggregation. To see all results, call .get_results(aggregated=False)
> FINALIZE phase done.
>           Stored results at ~~
>           Overall runtime of this TimeEval run: 0.06137847900390625 seconds
>         
> Empty DataFrame
> Columns: [ROC_AUC_mean, FIXED_RANGE_PR_AUC_mean, train_main_time_mean, execute_main_time_mean, repetitions]
> Index: []

I thought error is occured at this part.
https://github.com/HPI-Information-Systems/TimeEval/blob/0a45790cbeeb55d6140a250c672d4a93d152dfa4/timeeval/core/times.py#L31

and I wonder why the output is `Empty DataFrame`.

This is Dataset information.
![image](https://user-images.githubusercontent.com/99858033/205017073-6a528f2d-7584-4c7a-84cf-f30670444363.png)

**Comment by wenig** [comment]
Hi @JeonghwaYoo-R,

You have written the following `HPI_CLUSTER.akita_dataset_paths[HPI_CLUSTER.BENCHMARK]`. Please exchange with `"."` and try again. You are using windows, but the [akita_dataset_paths](https://github.com/HPI-Information-Systems/TimeEval/blob/main/timeeval/constants.py#L65) is a linux style path and shouldn't be used in your case.

**Reply by JeonghwaYoo-R** [reply]
@wenig Is it right?  `dm = DatasetManager(".", custom_datasets_file=custom_datasets_path)`. 

I tried again by modifying the code, but the same problem occurred on Windows. 

The following error occurred even when the code was executed in a Linux environment.

> Running PREPARE phase
> Running EVALUATION phase
> Evaluating:   0%|                                              | 0/1 [00:00<?, ?it/s]Exception occurred during the evaluation of LOF on the dataset Dataset(datasetId=('custom', 'dataset_name'), dataset_type='unknown', training_type=<TrainingType.UNSUPERVISED: 'unsupervised'>, length=172800, dimensions=1, contamination=0.004548611111111111, min_anomaly_length=1, median_anomaly_length=12, max_anomaly_length=68, period_size=None, num_anomalies=37).
> Traceback (most recent call last):
>   File "/opt/conda/lib/python3.7/site-packages/urllib3/connectionpool.py", line 706, in urlopen
>     chunked=chunked,
>   File "/opt/conda/lib/python3.7/site-packages/urllib3/connectionpool.py", line 394, in _make_request
>     conn.request(method, url, **httplib_request_kw)
>   File "/opt/conda/lib/python3.7/http/client.py", line 1281, in request
>     self._send_request(method, url, body, headers, encode_chunked)
>   File "/opt/conda/lib/python3.7/http/client.py", line 1327, in _send_request
>     self.endheaders(body, encode_chunked=encode_chunked)
>   File "/opt/conda/lib/python3.7/http/client.py", line 1276, in endheaders
>     self._send_output(message_body, encode_chunked=encode_chunked)
>   File "/opt/conda/lib/python3.7/http/client.py", line 1036, in _send_output
>     self.send(msg)
>   File "/opt/conda/lib/python3.7/http/client.py", line 976, in send
>     self.connect()
>   File "/opt/conda/lib/python3.7/site-packages/docker/transport/unixconn.py", line 43, in connect
>     sock.connect(self.unix_socket)
> FileNotFoundError: [Errno 2] No such file or directory
> 
> During handling of the above exception, another exception occurred:
> 
> Traceback (most recent call last):
>   File "/opt/conda/lib/python3.7/site-packages/requests/adapters.py", line 449, in send
>     timeout=timeout
>   File "/opt/conda/lib/python3.7/site-packages/urllib3/connectionpool.py", line 756, in urlopen
>     method, url, error=e, _pool=self, _stacktrace=sys.exc_info()[2]
>   File "/opt/conda/lib/python3.7/site-packages/urllib3/util/retry.py", line 532, in increment
>     raise six.reraise(type(error), error, _stacktrace)
>   File "/opt/conda/lib/python3.7/site-packages/urllib3/packages/six.py", line 769, in reraise
>     raise value.with_traceback(tb)
>   File "/opt/conda/lib/python3.7/site-packages/urllib3/connectionpool.py", line 706, in urlopen
>     chunked=chunked,
>   File "/opt/conda/lib/python3.7/site-packages/urllib3/connectionpool.py", line 394, in _make_request
>     conn.request(method, url, **httplib_request_kw)
>   File "/opt/conda/lib/python3.7/http/client.py", line 1281, in request
>     self._send_request(method, url, body, headers, encode_chunked)
>   File "/opt/conda/lib/python3.7/http/client.py", line 1327, in _send_request
>     self.endheaders(body, encode_chunked=encode_chunked)
>   File "/opt/conda/lib/python3.7/http/client.py", line 1276, in endheaders
>     self._send_output(message_body, encode_chunked=encode_chunked)
>   File "/opt/conda/lib/python3.7/http/client.py", line 1036, in _send_output
>     self.send(msg)
>   File "/opt/conda/lib/python3.7/http/client.py", line 976, in send
>     self.connect()
>   File "/opt/conda/lib/python3.7/site-packages/docker/transport/unixconn.py", line 43, in connect
>     sock.connect(self.unix_socket)
> urllib3.exceptions.ProtocolError: ('Connection aborted.', FileNotFoundError(2, 'No such file or directory'))
> 
> During handling of the above exception, another exception occurred:
> 
> Traceback (most recent call last):
>   File "/opt/conda/lib/python3.7/site-packages/docker/api/client.py", line 214, in _retrieve_server_version
>     return self.version(api_version=False)["ApiVersion"]
>   File "/opt/conda/lib/python3.7/site-packages/docker/api/daemon.py", line 181, in version
>     return self._result(self._get(url), json=True)
>   File "/opt/conda/lib/python3.7/site-packages/docker/utils/decorators.py", line 46, in inner
>     return f(self, *args, **kwargs)
>   File "/opt/conda/lib/python3.7/site-packages/docker/api/client.py", line 237, in _get
>     return self.get(url, **self._set_request_timeout(kwargs))
>   File "/opt/conda/lib/python3.7/site-packages/requests/sessions.py", line 555, in get
>     return self.request('GET', url, **kwargs)
>   File "/opt/conda/lib/python3.7/site-packages/requests/sessions.py", line 542, in request
>     resp = self.send(prep, **send_kwargs)
>   File "/opt/conda/lib/python3.7/site-packages/requests/sessions.py", line 655, in send
>     r = adapter.send(request, **kwargs)
>   File "/opt/conda/lib/python3.7/site-packages/requests/adapters.py", line 498, in send
>     raise ConnectionError(err, request=request)
> requests.exceptions.ConnectionError: ('Connection aborted.', FileNotFoundError(2, 'No such file or directory'))
> 
> During handling of the above exception, another exception occurred:
> 
> Traceback (most recent call last):
>   File "/opt/conda/lib/python3.7/site-packages/timeeval/timeeval.py", line 326, in _run
>     result = exp.evaluate()
>   File "/opt/conda/lib/python3.7/site-packages/timeeval/core/experiments.py", line 93, in evaluate
>     y_scores, execution_times = self._perform_execution()
>   File "/opt/conda/lib/python3.7/site-packages/timeeval/core/experiments.py", line 163, in _perform_execution
>     y_scores, times = Times.from_execute_algorithm(self.algorithm, X, self.build_args())
>   File "/opt/conda/lib/python3.7/site-packages/timeeval/core/times.py", line 31, in from_execute_algorithm
>     x, main_time = timer(algorithm.execute, x, args)
>   File "/opt/conda/lib/python3.7/site-packages/timeeval/core/times.py", line 44, in timer
>     fn_result = fn(X, args)
>   File "/opt/conda/lib/python3.7/site-packages/timeeval/algorithm.py", line 151, in execute
>     return self.main(dataset, args)
>   File "/opt/conda/lib/python3.7/site-packages/timeeval/adapters/base.py", line 17, in __call__
>     return self._call(dataset, args)
>   File "/opt/conda/lib/python3.7/site-packages/timeeval/adapters/docker.py", line 213, in _call
>     container = self._run_container(dataset, args)
>   File "/opt/conda/lib/python3.7/site-packages/timeeval/adapters/docker.py", line 118, in _run_container
>     client = docker.from_env()
>   File "/opt/conda/lib/python3.7/site-packages/docker/client.py", line 101, in from_env
>     **kwargs_from_env(**kwargs)
>   File "/opt/conda/lib/python3.7/site-packages/docker/client.py", line 45, in __init__
>     self.api = APIClient(*args, **kwargs)
>   File "/opt/conda/lib/python3.7/site-packages/docker/api/client.py", line 197, in __init__
>     self._version = self._retrieve_server_version()
>   File "/opt/conda/lib/python3.7/site-packages/docker/api/client.py", line 222, in _retrieve_server_version
>     'Error while fetching server API version: {0}'.format(e)
> docker.errors.DockerException: Error while fetching server API version: ('Connection aborted.', FileNotFoundError(2, 'No such file or directory'))
> Evaluating: 100%|██████████████████████████████████████| 1/1 [00:00<00:00, 23.62it/s]
> Running FINALIZE phase
> Traceback (most recent call last):
>   File "/opt/conda/lib/python3.7/site-packages/urllib3/connectionpool.py", line 706, in urlopen
>     chunked=chunked,
>   File "/opt/conda/lib/python3.7/site-packages/urllib3/connectionpool.py", line 394, in _make_request
>     conn.request(method, url, **httplib_request_kw)
>   File "/opt/conda/lib/python3.7/http/client.py", line 1281, in request
>     self._send_request(method, url, body, headers, encode_chunked)
>   File "/opt/conda/lib/python3.7/http/client.py", line 1327, in _send_request
>     self.endheaders(body, encode_chunked=encode_chunked)
>   File "/opt/conda/lib/python3.7/http/client.py", line 1276, in endheaders
>     self._send_output(message_body, encode_chunked=encode_chunked)
>   File "/opt/conda/lib/python3.7/http/client.py", line 1036, in _send_output
>     self.send(msg)
>   File "/opt/conda/lib/python3.7/http/client.py", line 976, in send
>     self.connect()
>   File "/opt/conda/lib/python3.7/site-packages/docker/transport/unixconn.py", line 43, in connect
>     sock.connect(self.unix_socket)
> FileNotFoundError: [Errno 2] No such file or directory
> 
> During handling of the above exception, another exception occurred:
> 
> Traceback (most recent call last):
>   File "/opt/conda/lib/python3.7/site-packages/requests/adapters.py", line 449, in send
>     timeout=timeout
>   File "/opt/conda/lib/python3.7/site-packages/urllib3/connectionpool.py", line 756, in urlopen
>     method, url, error=e, _pool=self, _stacktrace=sys.exc_info()[2]
>   File "/opt/conda/lib/python3.7/site-packages/urllib3/util/retry.py", line 532, in increment
>     raise six.reraise(type(error), error, _stacktrace)
>   File "/opt/conda/lib/python3.7/site-packages/urllib3/packages/six.py", line 769, in reraise
>     raise value.with_traceback(tb)
>   File "/opt/conda/lib/python3.7/site-packages/urllib3/connectionpool.py", line 706, in urlopen
>     chunked=chunked,
>   File "/opt/conda/lib/python3.7/site-packages/urllib3/connectionpool.py", line 394, in _make_request
>     conn.request(method, url, **httplib_request_kw)
>   File "/opt/conda/lib/python3.7/http/client.py", line 1281, in request
>     self._send_request(method, url, body, headers, encode_chunked)
>   File "/opt/conda/lib/python3.7/http/client.py", line 1327, in _send_request
>     self.endheaders(body, encode_chunked=encode_chunked)
>   File "/opt/conda/lib/python3.7/http/client.py", line 1276, in endheaders
>     self._send_output(message_body, encode_chunked=encode_chunked)
>   File "/opt/conda/lib/python3.7/http/client.py", line 1036, in _send_output
>     self.send(msg)
>   File "/opt/conda/lib/python3.7/http/client.py", line 976, in send
>     self.connect()
>   File "/opt/conda/lib/python3.7/site-packages/docker/transport/unixconn.py", line 43, in connect
>     sock.connect(self.unix_socket)
> urllib3.exceptions.ProtocolError: ('Connection aborted.', FileNotFoundError(2, 'No such file or directory'))
> 
> During handling of the above exception, another exception occurred:
> 
> Traceback (most recent call last):
>   File "/opt/conda/lib/python3.7/site-packages/docker/api/client.py", line 214, in _retrieve_server_version
>     return self.version(api_version=False)["ApiVersion"]
>   File "/opt/conda/lib/python3.7/site-packages/docker/api/daemon.py", line 181, in version
>     return self._result(self._get(url), json=True)
>   File "/opt/conda/lib/python3.7/site-packages/docker/utils/decorators.py", line 46, in inner
>     return f(self, *args, **kwargs)
>   File "/opt/conda/lib/python3.7/site-packages/docker/api/client.py", line 237, in _get
>     return self.get(url, **self._set_request_timeout(kwargs))
>   File "/opt/conda/lib/python3.7/site-packages/requests/sessions.py", line 555, in get
>     return self.request('GET', url, **kwargs)
>   File "/opt/conda/lib/python3.7/site-packages/requests/sessions.py", line 542, in request
>     resp = self.send(prep, **send_kwargs)
>   File "/opt/conda/lib/python3.7/site-packages/requests/sessions.py", line 655, in send
>     r = adapter.send(request, **kwargs)
>   File "/opt/conda/lib/python3.7/site-packages/requests/adapters.py", line 498, in send
>     raise ConnectionError(err, request=request)
> requests.exceptions.ConnectionError: ('Connection aborted.', FileNotFoundError(2, 'No such file or directory'))
> 
> During handling of the above exception, another exception occurred:
> 
> Traceback (most recent call last):
>   File "timeeval_lof.py", line 39, in <module>
>     timeeval.run()
>   File "/opt/conda/lib/python3.7/site-packages/timeeval/timeeval.py", line 594, in run
>     self._finalize()
>   File "/opt/conda/lib/python3.7/site-packages/timeeval/timeeval.py", line 522, in _finalize
>     algorithm.finalize()
>   File "/opt/conda/lib/python3.7/site-packages/timeeval/algorithm.py", line 195, in finalize
>     fn()
>   File "/opt/conda/lib/python3.7/site-packages/timeeval/adapters/docker.py", line 237, in finalize
>     client = docker.from_env(timeout=Duration("10 minutes").to_seconds())
>   File "/opt/conda/lib/python3.7/site-packages/docker/client.py", line 101, in from_env
>     **kwargs_from_env(**kwargs)
>   File "/opt/conda/lib/python3.7/site-packages/docker/client.py", line 45, in __init__
>     self.api = APIClient(*args, **kwargs)
>   File "/opt/conda/lib/python3.7/site-packages/docker/api/client.py", line 197, in __init__
>     self._version = self._retrieve_server_version()
>   File "/opt/conda/lib/python3.7/site-packages/docker/api/client.py", line 222, in _retrieve_server_version
>     'Error while fetching server API version: {0}'.format(e)
> docker.errors.DockerException: Error while fetching server API version: ('Connection aborted.', FileNotFoundError(2, 'No such file or directory'))

**Comment by SebastianSchmidl** [comment]
Hi @JeonghwaYoo-R,

Your error seems very strange, and we can't reproduce it. Please provide a complete [MRE](https://stackoverflow.com/help/minimal-reproducible-example). We need:

- your Python script/code
- the dataset and configuration files (and their names)
- the TimeEval version (`python -c "import timeeval; print(timeeval.__version__)"`)
- anything else that your code needs to run

When you have built your MRE, please test the following:

**A1**: `FunctionAdapter` on Linux
**A2**: `FunctionAdapter` on Windows
**B1**: `DockerAdapter` on Linux
**B2**: `DockerAdapter` on Windows

> **Please first test on Linux!** If it works on Linux, we can already narrow it down to be a problem with Windows.

Please try to run your example without using the `DockerAdapter`. If it works without it, there is an issue with Docker. You can try the following demo-function:

```python
from timeeval import Algorithm
from timeeval.adapters import FunctionAdapter
from timeeval.data_types import AlgorithmParameter
import numpy as np


def algorithm_function(data: AlgorithmParameter, args: dict) -> np.ndarray:
    if isinstance(data, np.ndarray):
        return np.zeros_like(data)
    else:  # data = pathlib.Path
        return np.genfromtxt(data)[0]

algorithms = [
    Algorithm(
        name="MyPythonFunctionAlgorithm",
        main=FunctionAdapter(algorithm_function),
        data_as_file=False
    )
]
```

---
> and I wonder why the output is Empty DataFrame.

This is expected because `aggregate=True` is the default in [`get_results()`](https://github.com/HPI-Information-Systems/TimeEval/blob/main/timeeval/timeeval.py#L390) and this filters out all failing experiments. In your case, no successful experiments are left in the result DataFrame. The warning in the stdout of TimeEval tells you exactly that:

> ```
> [...]
> The results contain errors which are filtered out for the final aggregation. To see all results, call .get_results(aggregated=False)
> [...]
> ```
So, you can try `get_results(aggregate=False)` to access the complete results. Alternatively, you can inspect the results in the `results.csv`-File (per default: `./results/results.csv`)

**Reply by JeonghwaYoo-R** [reply]
@CodeLionX 
Thank you for your reply!

> * your Python script/code

My python code is [here](https://github.com/HPI-Information-Systems/TimeEval/discussions/53#discussion-4617969)

> * the dataset and configuration files (and their names)

My configuration file (json) is here. 
```
{
  "dataset_name": {
    "test_path": "C:/Dev/Python/anomalydetector/samples/gentle_sensor_raw_data_trial2_uncertain_0_normalize_1s_timeeval.csv"
  }
}
```

Dataset (`gentle_sensor_raw_data_trial2_uncertain_0_normalize_1s_timeeval.csv`) is here. 
```
timestamp,value,is_anomaly
2022-05-24 00:00:00,1.47,0.0
2022-05-24 00:00:01,1.47,0.0
2022-05-24 00:00:02,1.26,0.0
2022-05-24 00:00:03,1.26,0.0
~~~
```

> * the TimeEval version (`python -c "import timeeval; print(timeeval.__version__)"`)

1.2.6

> * anything else that your code needs to run

docer version is 4.4.4

> A1: FunctionAdapter on Linux

-> Working (The value of status is Status.OK in results.csv)

> A2: FunctionAdapter on Windows

-> Working (The value of status is Status.OK in results.csv)

> B1: DockerAdapter on Linux

-> [Error](https://github.com/HPI-Information-Systems/TimeEval/discussions/53#discussioncomment-4318759)

> B2: DockerAdapter on Windows

-> [Error ](https://github.com/HPI-Information-Systems/TimeEval/discussions/53#discussion-4617969)

**Reply by SebastianSchmidl** [reply]
Ok, this cleared some things up:

1. I doubt that it's a problem with your custom datasets.
2. It just fails using the `DockerAdapter`, so it's an issue with Docker.
3. It's very strange that the error message is the same on your Linux and Windows hosts. Can you confirm that again?

I suspect that TimeEval can't find your Docker-installation, especially because of these lines in your error message:

> File "/opt/conda/lib/python3.7/site-packages/docker/transport/unixconn.py", line 43, in connect
> sock.connect(self.unix_socket)
> FileNotFoundError: [Errno 2] No such file or directory

---

Let's first **focus on your Linux machine** because I'm more familiar with that. Please provide the following details:

1. I assume that 

   > docer version is 4.4.4

   is the version of the python docker package (`pip freeze | grep docker`). Is that correct?

2. What is the version of your docker runtime system? You can use the docker client (Linux CLI) to get the version numbers of the server and client: `docker version`.  Please add the complete output.

3. Per default, TimeEval and the docker Python package uses the docker-socket to connect to it. Can you confirm that it's there? `ls -alh /var/run/docker.sock`

4. Maybe your environment variables are compromised or set to something different. Please check the Docker-relevant variables and add them to your reply:

   ```bash
   echo $DOCKER_HOST
   echo $DOCKER_CERT_PATH
   echo $DOCKER_TLS_VERIFY
   ```

5. Please run the following script and upload the output:

```python
#!/usr/bin/env python3
import json
import docker


def main():
    client = docker.from_env()
    client.ping()

    version_info = client.version()
    print(json.dumps(version_info, indent=2))


if __name__ == "__main__":
    main()
```
